### PR TITLE
WIP: Fix segfault in QRCodeDetector.detectAndDecodeCurved

### PR DESCRIPTION
**Pull request by AleksandrPanov**
fixes https://github.com/opencv/opencv/issues/22892
merge with https://github.com/opencv/opencv_extra/pull/1072

The issue is in spline interpolation. Only the points of the function can be interpolated using splines, but the contour points may not be a function. There may be several points with the same x coordinate and different y. It results in getting NaN when calculating spline.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/objdetect/src/qrcode.cpp [hunk 1]
```diff
@@ -1973,6 +1973,29 @@ vector<vector<float> > QRDecode::computeSpline(const vector<int> &x_arr, const v
...
+    }
+    if (countX != 0 || countY != 0) {
+        vector<int>& newX = countX < countY ? x_arr : y_arr;
+        const int delta = newX.back() > newX.front() ? 1 : -1;
+        horizontal_order = countX < countY ? true : false;
+        if (min(countX, countY) > 0)
+            for (size_t i = 1ull; i < x_arr.size(); i++) {
+                if (newX[i] == newX[i-1])
+                    newX[i] += delta;
+            }
```
Comment (asmorkalov) [inline modules/objdetect/src/qrcode.cpp:1994]: This hack requires description right in the code. Looks like some array values are modified on +-1, but the reason and idea is not obvious.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

File: modules/objdetect/src/qrcode.cpp (hunk 1 above)
Comment (AleksandrPanov) [reply modules/objdetect/src/qrcode.cpp:1994]: Added comment

```
Only the points of the function can be interpolated using splines.
Contour points may not be a function, there may be several points with the same x coordinate and different y.
We are adding a small offset in x coord to fix this problem. This should not significantly affect the final answer.
```

**Comment by AleksandrPanov** [comment]
The problem was in the original algorithm - PR #18003

@APrigarina, could you look at this solution or suggest another one?

**Review by asmorkalov — APPROVED** [review]

**Comment by vpisarev** [comment]
@AleksandrPanov, @asmorkalov, frankly speaking, I don't like this hack. There is a general flexible method how to interpolate any polyline or any polygon using a spline. First, you introduce "time" axis t. The simplest (and probably good enough for this purpose) method is to set t=0 for the first point, t=1 for the second point etc. Then, instead of computing spline as y=y(x) you just build splines x=x(t) and y=y(t). Then you can compute intermediate points at any precision by choosing some fractional t values. In this case you don't need the proposed hack, because you can be sure that all t's are different.

**Comment by asmorkalov** [comment]
thanks a lot for the parametrization idea!

File: modules/objdetect/src/qrcode.cpp [hunk 2]
```diff
@@ -1973,10 +1973,48 @@ vector<vector<float> > QRDecode::computeSpline(const vector<int> &x_arr, const v
     return S;
 }
 
-bool QRDecode::createSpline(vector<vector<Point2f> > &spline_lines)
+enum SplinePointsStatus {
+    OK = 0,
+    NEED_SWAP_AXES,
+    NEED_ADD_PARAMETRIZATION
```
Comment (asmorkalov) [inline modules/objdetect/src/qrcode.cpp:1979]: With parameterized spline we do not need branches any more.

**Review by asmorkalov — CHANGES_REQUESTED** [review]

**Comment by asmorkalov** [comment]
Replaced by https://github.com/opencv/opencv/pull/25026

**Pull request closed without merging**